File: typescript/src/types/binary-serializer-enum.ts
import { DataReader } from '../buffers/data-reader.interface';
import { DataWriter } from '../buffers/data-writer.interface';
import { BinaryTypeSerializer } from './binary-type-serializer.interface';
import Enum from '@barchart/common-js/lang/Enum';

/**
 *  Reads (and writes) enumeration values to (and from) a binary data source.
 *
 * @public
 * @exported
 * @implements {BinaryTypeSerializer<T>}
 * @template T - The enumeration type.
 */
export class BinarySerializerEnum<T extends Enum> implements BinaryTypeSerializer<T> {
    get sizeInBytes(): number {
        return 4;
    }

    private readonly binarySerializerInt: BinaryTypeSerializer<number>;
    private readonly enumType: new (...args: any[]) => T;

    constructor(binarySerializerInt: BinaryTypeSerializer<number>, enumType: new (...args: any[]) => T) {
        this.binarySerializerInt = binarySerializerInt;
        this.enumType = enumType;
    }

    encode(writer: DataWriter, value: T): void {
        this.binarySerializerInt.encode(writer, value.mapping as number);
    }

    decode(reader: DataReader): T {
        const intValue = this.binarySerializerInt.decode(reader);
        return Enum.fromMapping(this.enumType, intValue) as T;
    }

    getEquals(a: T, b: T): boolean {
        return this.binarySerializerInt.getEquals(a.mapping as number, b.mapping as number);
    }
}

File: dotnet/Barchart.BinarySerializer/Types/BinarySerializerEnum.cs
﻿#region Using Statements

using Barchart.BinarySerializer.Buffers;

#endregion

namespace Barchart.BinarySerializer.Types;

/// <summary>
///     Reads (and writes) enumeration values to (and from) a binary data source.
///     Serialization uses a 32-bit integer (with a maximum of 2.1 billion items).
///     In other words, this serializer probably uses too much bandwidth considering
///     that most enumerations contain relatively few items.
/// </summary>
/// <typeparam name="T">
///     The enumeration type.
/// </typeparam>
public class BinarySerializerEnum<T> : IBinaryTypeSerializer<T> where T : Enum
{
    #region Fields

    private readonly IBinaryTypeSerializer<int> _binarySerializerInt;

    #endregion

    #region Constructor(s)

    /// <summary>
    ///     Create an instance of <see cref="BinarySerializerEnum{T}"/>.
    /// </summary>
    /// <param name="binarySerializerInt">
    ///     The binary serializer for 32-bit integers.
    /// </param>
    public BinarySerializerEnum(IBinaryTypeSerializer<int> binarySerializerInt)
    {
        _binarySerializerInt = binarySerializerInt;
    }

    #endregion

    #region Methods

    /// <inheritdoc />
    public void Encode(IDataBufferWriter writer, T value)
    {
        _binarySerializerInt.Encode(writer, Convert.ToInt32(value));
    }

    /// <inheritdoc />
    public T Decode(IDataBufferReader reader)
    {
        return (T)Enum.Parse(typeof(T), _binarySerializerInt.Decode(reader).ToString(), true);
    }

    /// <inheritdoc />
    public bool GetEquals(T a, T b)
    {
        return _binarySerializerInt.GetEquals(Convert.ToInt32(a), Convert.ToInt32(b));
    }

    #endregion
}